code/modules/admin/banjob.dm

```diff
@@ -1,159 +1,62 @@
-	jobban_remove("[ckey] - [rank]")
+	while(query.NextRow())
+		jobs += query.item[1]
 
-/proc/jobban_remove(X)
-	for(var/i = 1; i <= length(GLOB.jobban_keylist); i++)
-		if( findtext(GLOB.jobban_keylist[i], "[X]") )
-			// This need to be here, instead of jobban_unban, due to direct calls to jobban_remove
-			if(GLOB.jobban_regex.Find(X))
-				var/ckey = GLOB.jobban_regex.group[1]
-				var/rank = GLOB.jobban_regex.group[2]
-				if(GLOB.jobban_assoclist[ckey] && GLOB.jobban_assoclist[ckey][rank])
-					GLOB.jobban_assoclist[ckey] -= rank
-				else
-					log_runtime(EXCEPTION("Attempted to remove non-existent job ban: [X]"))
-			else
-				log_runtime(EXCEPTION("Failed to remove malformed job ban from associative list: [X]"))
-			GLOB.jobban_keylist.Remove(GLOB.jobban_keylist[i])
-			return 1
-	return 0
+	return jobs
 
-/mob/verb/displayjobbans()
+/client/verb/displayjobbans()
 	set category = "OOC"
 	set name = "Display Current Jobbans"
 	set desc = "Displays all of your current jobbans."
 
-	if(!client || !ckey)
-		return
+	// Ok. I know this verb here is scoped to client, I know.
+	// But sometimes when executing, the src will be a mob
+	// I have no idea why, but this is a workaround.
+	jbh.reload_jobbans(usr.client)
 
 	var/is_actually_banned = FALSE
-	var/datum/db_query/select_query = SSdbcore.NewQuery({"
-		SELECT bantime, bantype, reason, job, duration, expiration_time, a_ckey FROM ban
-		WHERE ckey LIKE :ckey AND ((bantype like 'JOB_TEMPBAN' AND expiration_time > Now()) OR (bantype like 'JOB_PERMABAN')) AND isnull(unbanned)
-		ORDER BY bantime DESC LIMIT 100"},
-		list("ckey" = ckey)
-	)
-
-	if(!select_query.warn_execute())
-		qdel(select_query)
-		return FALSE
-
-	while(select_query.NextRow())
-
-		var/bantime = select_query.item[1]
-		var/bantype  = select_query.item[2]
-		var/reason = select_query.item[3]
-		var/job = select_query.item[4]
-		var/duration = select_query.item[5]
-		var/expiration = select_query.item[6]
-		var/ackey = select_query.item[7]
+	if(length(jbh.job_bans))
+		is_actually_banned = TRUE
 
-		if(bantype == "JOB_PERMABAN")
-			to_chat(src, "<span class='warning'>[bantype]: [job] - REASON: [reason], by [ackey]; [bantime]</span>")
-		else if(bantype == "JOB_TEMPBAN")
-			to_chat(src, "<span class='warning'>[bantype]: [job] - REASON: [reason], by [ackey]; [bantime]; [duration]; expires [expiration]</span>")
+	if(!is_actually_banned)
+		to_chat(src, "<span class='warning'>You have no active jobbans!</span>")
+		return
 
-		is_actually_banned = TRUE
+	for(var/ban in jbh.job_bans)
+		var/datum/job_ban/JB = jbh.job_bans[ban] // Remember. Its assoc.
+		if(JB.bantype == "JOB_PERMABAN")
+			to_chat(src, "<span class='warning'>[JB.bantype]: [JB.job] - REASON: [JB.reason], by [JB.a_ckey]; [JB.bantime]</span>")
+		else if(JB.bantype == "JOB_TEMPBAN")
+			to_chat(src, "<span class='warning'>[JB.bantype]: [JB.job] - REASON: [JB.reason], by [JB.a_ckey]; [JB.bantime]; [JB.duration]; expires [JB.expiration_time]</span>")
```

Comment:
Could probably be changed to `switch(JB.bantype)` (bonus points for `#define`s)

Reply:
Will do

Reply:
The defines are numerical
```dm
#define BANTYPE_PERMA		1
#define BANTYPE_TEMP		2
#define BANTYPE_JOB_PERMA	3
#define BANTYPE_JOB_TEMP	4
#define BANTYPE_ANY_FULLBAN	5 //used to locate stuff to unban.
#define BANTYPE_ADMIN_PERMA	7
#define BANTYPE_ADMIN_TEMP	8
```